src/main/java/net/minecraftforge/common/PlantType.java

```diff
@@ -19,27 +19,71 @@
+    public static final PlantType PLAINS = get("plains", x -> x.getState().is(Blocks.GRASS_BLOCK) || Tags.Blocks.DIRT.contains(x.getState().getBlock()) || x.getState().is(Blocks.FARMLAND));
+    public static final PlantType DESERT = get("desert", x -> x.getState().is(Blocks.SAND) || x.getState().is(Blocks.TERRACOTTA) || x.getState().getBlock() instanceof GlazedTerracottaBlock);
+    public static final PlantType BEACH = get("beach", x -> {
+        boolean isBeach = x.getState().is(Blocks.GRASS_BLOCK) || net.minecraftforge.common.Tags.Blocks.DIRT.contains(x.getState().getBlock()) || x.getState().is(Blocks.SAND) || x.getState().is(Blocks.RED_SAND);
+        boolean hasWater = false;
+        for (Direction face : Direction.Plane.HORIZONTAL) {
+            BlockState blockState = x.getWorld().getBlockState(x.getPos().relative(face));
+            net.minecraft.fluid.FluidState fluidState = x.getWorld().getFluidState(x.getPos().relative(face));
+            hasWater |= blockState.is(Blocks.FROSTED_ICE);
+            hasWater |= fluidState.is(net.minecraft.tags.FluidTags.WATER);
+            if (hasWater)
+                break; //No point continuing.
+        }
+        return isBeach && hasWater;
+    });
+    public static final PlantType CAVE = get("cave", x -> x.getState().isFaceSturdy(x.getWorld(), x.getPos(), Direction.UP));
+    public static final PlantType WATER = get("water", x -> x.getState().getMaterial() == Material.WATER);
+    public static final PlantType NETHER_WART = get("nether_wart", x -> x.getState().is(Blocks.SOUL_SAND));
+    public static final PlantType CROP = get("crop", x -> x.getState().is(Blocks.FARMLAND));
+    public static final PlantType CACTUS = get("cactus", x -> x.getState().is(Blocks.CACTUS) || x.getState().is(Blocks.SAND) || x.getState().is(Blocks.RED_SAND));
+    public static final PlantType SUGAR_CANE = get("sugar_cane", BEACH.getBlockPredicate().or(x -> x.getState().is(Blocks.SUGAR_CANE)));
+    public static final PlantType NYLIUM_PLANTS = get("nylium_plants", PLAINS.getBlockPredicate().or(x -> x.getState().is(BlockTags.NYLIUM) || x.getState().is(Blocks.SOUL_SOIL)));
+    public static final PlantType SEA_PICKLE = get("sea_pickle", x -> {
+        BlockState below = x.getWorld().getBlockState(x.getPos().below());
+        return !below.getCollisionShape(x.getWorld(), x.getPos()).getFaceShape(Direction.UP).isEmpty() || below.isFaceSturdy(x.getWorld(), x.getPos(), Direction.UP);
+    });
+    public static final PlantType WITHER_ROSE = get("wither_rose", PLAINS.getBlockPredicate().or(x -> x.getState().is(Blocks.NETHERRACK) || x.getState().is(Blocks.SOUL_SAND) || x.getState().is(Blocks.SOUL_SOIL)));
+    public static final PlantType FUNGUS = get("fungus", PLAINS.getBlockPredicate().or(x -> x.getState().is(BlockTags.NYLIUM) || x.getState().is(Blocks.MYCELIUM) || x.getState().is(Blocks.SOUL_SOIL)));
+    public static final PlantType DEAD_BUSH = get("dead_bush", x ->
+    {
```

Comment:
Please choose a consistent formatting for the lambda brackets. I personally think they are more confusing on new lines, but at least do the same for all of them.